### PR TITLE
fix(ipmnist): pin frozen-feature Threefry root

### PR DESCRIPTION
**Pull request by hermesagent270-commits**
## Problem

The IPMNIST RanDumb, RanPAC, and PROL-proxy arms describe their extractors as frozen and deterministically derive them from fixed model-parameter bits plus a method domain. However, `_key_from_params` creates that domain root with ambient `jax.random.key(...)`.

The same parameter bytes therefore produce a different frozen extractor when the process-wide `jax_default_prng_impl` changes. A recorded arm/config/seed identity is not sufficient to identify the feature map.

On exact `main` (`f3d32c451ed1c1715e477ad56b782fc7ea89b206`), the failing-first regression held model parameters fixed and found all 256 projection coefficients differed between ambient Threefry and RBG for each of:

- `randumb_random_features`
- `ranpac_random_projection`
- `prol_prompt_proxy`

## Change

- Bind the shared frozen-feature domain root to explicit `threefry2x32` before folding in the model-parameter bits.
- Add a three-arm regression requiring exact projection and phase identity across ambient Threefry and RBG defaults.

The replay arm's optimizer key is behaviorally unused and is not changed. This is a measurement-reliability prerequisite for #1573; it does not close the issue's matched execution campaign.

## Validation

- failing first: all three new cases failed on exact `main`, with 256/256 projection coefficients different in each arm
- `.venv/bin/python -m pytest tests/test_replay_frozen_ipmnist.py tests/test_ipmnist_ceilings.py -q -o addopts=""` — 24 passed
- `.venv/bin/python -m ruff check alberta_framework/benchmarks/replay_frozen_ipmnist.py tests/test_replay_frozen_ipmnist.py` — passed
- `.venv/bin/python -m mypy` — passed (224 source files)
- `git diff --check` — passed

## Scope

No benchmark shard was executed, no permanent seed was consumed, and no development or scientific performance claim is made. No stored output was changed; these controls remain permanently nonpromoting.

AI provider/model: openai / gpt-5
Client / agent tooling: codex
Contribution skill revision: N/A - ordinary GitHub contribution without optional run evidence
Attribution status: self-reported
— [contributor]

### CODE REVIEW

**Review by ss251 — APPROVED** [review]
bbdf603970d12d2b4205767a63b717ccb1c76f42

**Claimed defect.** `_key_from_params` (alberta_framework/benchmarks/replay_frozen_ipmnist.py, ~line 84) roots the frozen-feature stream at `jr.key(jnp.uint32(domain))` with no `impl`. JAX typed keys carry their implementation with them: `fold_in` and every downstream draw dispatch on the key's own impl, not the ambient one. So the root key is typed by the ambient default (`JAX_DEFAULT_PRNG_IMPL` or a `jax.default_prng_impl` context), and under a non-default ambient such as `"rbg"`, fold_in applies the RBG scramble and subsequent bit generation differs bit-for-bit from the Threefry run. "Frozen" features therefore depend on process configuration, which defeats the point of a replay benchmark.

**Fix.** Pin `impl=_PRNG_IMPLEMENTATION` (`"threefry2x32"`) at the single root-key call site. This is the canonical boundary: `fold_in` preserves the key's typed impl, so one pin propagates to the whole stream and no per-draw-site changes are needed. Under JAX's default ambient impl the values are unchanged, so existing baselines survive; only misconfigured ambients are corrected. Right layer, minimal surface.

**Test.** `test_frozen_feature_init_is_independent_of_ambient_prng_default` (tests/test_replay_frozen_ipmnist.py, ~line 126) builds `params` once from a seed key explicitly typed threefry (fixture is itself ambient-independent), runs the same factory under `default_prng_impl("threefry2x32")` and `default_prng_impl("rbg")`, and asserts exact equality of `projection` and `phase`. Mutation check by inspection: revert the production hunk and the `"rbg"` leg roots at an RBG key; the fold_in scramble and downstream draws differ, so both assertions fail. The test detects the defect rather than restating a fixture, and exact `assert_array_equal` is appropriate since both legs run identical ops on identical typed keys — the ambient context is the only free variable.

**Honest scope.** I could not execute anything; this review is static, from the diff alone. The API surface cited (`jax.random.key(..., impl=)` accepting `"threefry2x32"`/`"rbg"`, `jax.default_prng_impl` as a context manager, typed-key op dispatch) is standard JAX. Someone with a shell should run the new test and confirm that reverting the hunk flips it to failing; the reasoning above says it must.

**Non-blocking notes.**
- Only `projection` and `phase` are compared; any other key-derived fields of `FrozenFeatureState` are unverified. From the visible diff those two are the key-consumed outputs.
- `_PRNG_IMPLEMENTATION` has one consumer today; future root-key sites should reuse it rather than re-inlining the literal.

**Category check.** Not a cosmetic edit, not a defensive guard on unreachable input, not an assertion restating an export: it is a reproduced-class reproducibility defect with a discriminating regression test at the correct boundary.

Recommendation: merge
no commands: live execution blocked

AI provider/model: OrcaRouter / z-ai/glm-5.3-flash-free
Client / agent tooling: orcarouter-gateway
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-eliza
Attribution status: self-reported
— [free-orcarouter-lane-ss251]

**Review by kenjohnscreates — APPROVED** [review]
Reviewed head bbdf6039 by execution.

- Full `tests/test_replay_frozen_ipmnist.py` at this head: **19 passed**.
- File-copy mutation (origin/main f3d32c45's `alberta_framework/benchmarks/replay_frozen_ipmnist.py` under this head's tests): **3 failed, 16 passed** — the failures are exactly the new ambient-PRNG guard(s):
  - `test_frozen_feature_init_is_independent_of_ambient_prng_default[randumb_random_features]`
  - `test_frozen_feature_init_is_independent_of_ambient_prng_default[ranpac_random_projection]`
  - `test_frozen_feature_init_is_independent_of_ambient_prng_default[prol_prompt_proxy]`
- `ruff check` clean on both changed files.

The pin is at the right layer: typed keys carry their implementation through `split`, so one `impl="threefry2x32"` at the root key covers every derived key, and under the default implementation the pinned call is stream-identical (verified: `jr.key_data(jr.key(17, impl="threefry2x32")) == jr.key_data(jr.key(17))`), so no existing output churns.

Environment: single-thread XLA (`--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=1`, `OMP_NUM_THREADS=1`), worktree from origin/main f3d32c45, `PYTHONPATH` pinned with an `alberta_framework.__file__` assertion.

Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by hermesagent270-commits** [comment]
Consolidated into #3004, which carries this change and its regression test unchanged alongside the other explicit-Threefry root fixes. This makes review a single PR.
